src/disposable/AsyncDisposer.ts

```diff
@@ -1,5 +1,26 @@
 import { IAsyncDisposable } from './types';
 
-export class AsyncDisposer implements IAsyncDisposable {
-  constructor(public dispose: () => void | Promise<void>) {}
+const nope = () => {
+  console.warn('Already disposed');
+};
+
+declare class AsyncDisposer implements IAsyncDisposable {
+  constructor(dispose: () => void | Promise<void>);
+  public dispose(): void | Promise<void>;
+  public disposed: boolean;
+}
+
+function AsyncDisposer(_dispose: () => void | Promise<void>): AsyncDisposer {
+  const self: AsyncDisposer = {
+    async dispose() {
```

Comment:
TBH, I'd go with
```ts
const self = Object.defineProperties({}, {
  dispose: {
     enumerable: true, // or false (default)?
     configurable: true,
     async value() {
      await _dispose();
      Object.defineProperty(self, 'value', {value: nope, enumerable: true });
     },
   },
  disposed: {
     enumerable: true,
     get disposed() {
         return self.dispose === nope;
     }
   }
})
```

At the moment, `dispose` as well as `disposed` might be overwritten.

Reply:
> At the moment, `dispose` as well as `disposed` might be overwritten.

Ah fudge. And my implementation of `AsyncDisposerSet` relies on that behavior. Sigh. Hang on for another major refactoring.

Reply:
Yeah. To be more correct I should have said that `disposed` might be reconfigured, not overwritten, since it's just a getter without a setter.
In any case, yeah, we need to address the issue with `dispose`.

Reply:
I've got something. But it is getting increasingly complicated. :D